Timeline page based on stored leaderboard.

PedersenThomas opened the pull request:
* It introduces a page with a list of recent activity based on the 10 songs that are stored for each difficulty in each song.
* Makes it the landing page.

PedersenThomas committed:
Initial commit for adding Timeline

PedersenThomas committed:
Removed the duplication of ago.

PedersenThomas committed:
Removed comment

PedersenThomas committed:
Made it the default page.

PedersenThomas committed:
Added a few comments for future improvements to the code.

hoovercj commented on src/components/kpi-group/index.tsx:
```diff
@@ -34,7 +34,7 @@ export class KpiGroup extends React.Component<KpiGroupProps> {
     private renderKpi = (kpi: KpiData, showBorder: boolean = true) => {
         return (
-            <div key={kpi.name} className={'kpi-group_stat-wrapper'}>
+            <div key={kpi.name + kpi.name + kpi.subvalue + kpi.inlineValue} className={'kpi-group_stat-wrapper'}>
```
Why `kpi.name + kpi.name`? Suggest de-duplicating it
```suggestion
            <div key={kpi.name + kpi.subvalue + kpi.inlineValue} className={'kpi-group_stat-wrapper'}>
```

hoovercj replied:
Can you elaborate a bit more on why this is necessary or why you think this is the best fix? The idea behind a key in react is so that if you have a list of items, and the value of one of the items changes, only that element needs to be re-rendered. Or if an item is removed from the list, react can remove the item whose key no longer appears in the list without comparing the actual content. So for example, if there was a list of scores for players A, B, and C, if the key was the name of the player, then when the scores changed from A: 100, B: 200, C: 300 to A: 100, B: 400, C: 300, the elements for A, B, and C would not need to be destroyed and recreated, the only change in the DOM would be changing the text for B to 400.

So in this case, I was simply using the "name" as the key because in most of my kpis, the name was the difficulty or the player or something, and each instance of the kpi in a card was for a different difficulty or a different player.

I can see now that this is flawed because in your use case, the kpis might have the same name.

But given my explanation above, do you think that combining the name, subvalue, and inline value is the best way to make a key?

In the example above, changing the score would change the key, so react wouldn't know to reuse the element it previously used for B. It _might_ reuse it, but that would be more by chance/convenience.

So is there something else that can be done within the kpi group to create a key that is tied to the "identity" of a kpi without being tied to the values which might change?

**Note:** In the grand scheme of things, this key is fine, I'm just trying to get both of us to think about better ways if we can :-)

hoovercj replied:
discussed offline: keys should come from outside

hoovercj commented on src/pages/timeline-list/index.tsx:
```diff
@@ -0,0 +1,142 @@
+        return result.sort((n1, n2) => n2.score.timestamp - n1.score.timestamp);
+    }
+
+    private FindAtTheTimeBest(scores: SongScore[], refScore: SongScore): SongScore | null {
```
If I understand correctly, this function returns the best score before the current score, assuming that all of the scores passed in are sorted in ascending order by timestamp.

So would a better function name be `GetPreviousHighScore`?

hoovercj replied:
```suggestion
    private GetPreviousHighScore(scores: SongScore[], currentHighScore: SongScore): SongScore | null {
```

hoovercj commented on src/pages/timeline-list/index.tsx:
```diff
@@ -0,0 +1,142 @@
+            } else {
+                if(item.score.score > item.AtTheTimeBest.score) {
+                    return `${item.score.playerName} beat the previous best player ${item.AtTheTimeBest.playerName} ${dateToTimeDifferenceInWords(date)}.`;
+                } else {
```
```typescript
return `${item.score.playerName} hit the leaderboard but ${item.AtTheTimeBest.playerName} has a better score ${dateToTimeDifferenceInWords(date)}.`;
```

Does this "pollute" the timeline with a lot of entries?

PedersenThomas commented on src/pages/timeline-list/index.tsx:
```diff
@@ -0,0 +1,142 @@
+
+            return {
+                title: titleArtistString(item.song.title, item.song.artist),
+                subtitle: item.song.artist || undefined,
```
Should have been Author, not artist.

hoovercj reviewed (changes requested):
Here is some feedback!

PedersenThomas commented on src/pages/timeline-list/index.tsx:
```diff
@@ -0,0 +1,142 @@
+        return result.sort((n1, n2) => n2.score.timestamp - n1.score.timestamp);
+    }
+
+    private FindAtTheTimeBest(scores: SongScore[], refScore: SongScore): SongScore | null {
```
```suggestion
    private GetBestScoreBefore(scores: SongScore[], timestamp: number): SongScore | null {
```

PedersenThomas commented on src/pages/timeline-list/index.tsx:
```diff
@@ -0,0 +1,142 @@
+        let previous: SongScore | null = null;
+        for (const score of scores) {
+            if (score.timestamp >= refScore.timestamp) {
+                return previous
```
find a better name than previous.

hoovercj committed:
Dedupe kpi.name

Co-Authored-By: PedersenThomas <cooltomme@gmail.com>

PedersenThomas committed:
Review comments + Links

Added Key to the Kpi component
Only scores that hits the top of the leaderboard creates a entry on the timeline page.
Prepared the work for adding links to players, but is blocked by PR #13.
